[mlir][gpu]Add GPUToXeVM lowering pipeline pass.

mshahneo opened the pull request:
It's the default GPU to XeVM lowering pipeline. It starts by lowering GPU code to the specified compilation target (default is fatbin), then lowers the host code.
If XeGPU ops are used, it expects the MLIR code to have XeGPU ops already embedded in gpu code.

mshahneo committed:
Add GPUToXeVM lowering pipeline pass.

It's the default GPU to XeVM lowering pipeline. It starts by lowering GPU
code to the specified compilation target (default is fatbin),
then lowers the host code.
If XeGPU ops are used, it expects the MLIR code to have
XeGPU ops already embedded in gpu code.

mshahneo committed:
Add a pass option to provide the XeGPU code level.

XeGPU allows worgroup, subgroup, and workitem level programming.
This options lets the pass manager know at which level the
XeGPU ops belong to.

mshahneo committed:
Add a workitem level gemm test for XeGPU.

mshahneo committed:
Fix a small logic.

mshahneo committed:
Add test cases for SG and WG.

Jianhui-Li commented on mlir/test/Integration/Dialect/XeGPU/SIMT/simple_gemm.mlir:
```diff
@@ -0,0 +1,121 @@
+// RUN: mlir-opt %s --gpu-lower-to-xevm-pipeline="xegpu-op-level=workitem" \
```
Please change the directory and pipeline parameter to "lane", instead of SIMT and workitem, to be consistent with XeGPU layout definition.

mshahneo replied:
Thanks a lot, Jianhui. Changed it now.

mshahneo committed:
Address review comments.

Change `workitem` to `lane`.

mshahneo committed:
Address review comments.

Rename `SIMT` test folder to `LANE`.

mshahneo committed:
Address review comments.

Remove some unnecessary headers.

mshahneo committed:
Address review comments.

mshahneo committed:
Add a missing dependency.
Skip tests if certain condition is not fulfilled.

mshahneo committed:
Address review comments.

Remove the explicit use of `convert-xevm-to-llvm` pass.
`convert-to-llvm` already uses this pattern.

mshahneo committed:
Address review comments.

mshahneo committed:
Address review comments.

mshahneo commented on mlir/test/Integration/Dialect/XeGPU/LANE/simple_gemm.mlir:
```diff
@@ -0,0 +1,121 @@
+    %memref_c = gpu.alloc  () : memref<256x256xf32>
+    gpu.memcpy %memref_c, %c : memref<256x256xf32>, memref<256x256xf32>
+    gpu.launch_func @kernel::@simple_gemm blocks in (%c32, %c16, %c1) threads in (%c16, %c1, %c1) args(%memref_a : memref<256x256xf16>, %memref_b : memref<256x256xf16>, %memref_c : memref<256x256xf32>)
+    gpu.wait // Wait for the kernel to finish.
```
Actually we do use async. Through the usage of `GpuAsyncRegionPass`. This pass adds async keyword to gpu ops based on dependency analysis. This pass is necessary to generate asynchronous chain. These chains essentially gets converted to `mgpuCreateStream()`.

mshahneo commented on mlir/lib/Dialect/GPU/Pipelines/GPUToXeVMPipeline.cpp:
```diff
@@ -0,0 +1,136 @@
+    pm.addPass(createGpuXeVMAttachTarget(xevmTargetOptions));
+  }
+  pm.addPass(createLowerAffinePass());
+  pm.addNestedPass<func::FuncOp>(createGpuAsyncRegionPass());
```
We actually need this, as explained above.

mshahneo committed:
Add `ConvertMathToXeVMPass` to the pipeline.

Jianhui-Li reviewed (approved):
LGTM

mshahneo committed:
Address review comments.

Change the utility functions names.

mshahneo committed:
Fix formatting.
